Add detailed list of which directories the longhorn components need (#458)
longhorn/website: content/docs/1.2.3/deploy/install/_index.md

### Root and Privileged Permission

Longhorn components require root access with privileged permissions to achieve volume operations and management, because Longhorn relies on system resources on the host across different namespaces, for example, Longhorn uses `nsenter` to understand block devices' usage or encrypt/decrypt volumes on the host.

> Below are the list of the directories of all components are:

Reviewer: ```suggestion
Below are the directories Longhorn components requiring access with root and privileged permissions :
```

- Longhorn Manager

>   - /dev: Longhorn is a block storage solution for Kubernetes. The block device is created under `/dev` path.

Reviewer: ```suggestion
  - /dev: Block devices created by Longhorn are under the `/dev` path.
```

>   - /proc: To access host's mount namespace by `nsenter` across different namepsaces.

Reviewer: ```suggestion
  - /proc: Find the recognized host process like container runtime, then use `nsenter` to access the mounts on the host to understand disks usage.
```

>   - /var/lib/longhorn: The path to use for storing data on a host.

Reviewer: ```suggestion
  - /var/lib/longhorn: The default path for storing volume data on a host.
```

> - Longhorn Engine

Reviewer: Do we need this? I believe this should be covered in IM engine instead.

Author: This should be Longhorn Engine Image DaemonSet, I'll rename it.

>   - /var/lib/longhorn/engine-binaries: The path where the Longhorn engine binaries located.

Reviewer: ```suggestion
  - /var/lib/longhorn/engine-binaries: The default path for storing the Longhorn engine binaries.
```

- Longhorn Instance Manager (Engine)

>   - /dev: Longhorn is a block storage solution for Kubernetes. The block device is created under `/dev` path.

Reviewer: ```suggestion
  - /dev: Block devices created by Longhorn are under the `/dev` path.
```

>   - /proc: To access host's mount namespace by `nsenter` across different namepsaces.

Reviewer: ```suggestion
  - /proc: Find the recognized host process like container runtime, then use `nsenter` to manage iSCSI targets and initiators, also some file system operations like sync before snapshotting. 
```

>   - /var/lib/longhorn/engine-binaries: The path where the Longhorn engine binaries located.

Reviewer: ```suggestion
  - /var/lib/longhorn/engine-binaries: The default path for storing the Longhorn engine binaries.
```

- Longhorn Instance Manager (Replica)
  - /: Data path (allow the user to customize data path)
- Longhorn Share Manager
  - /dev: Volume block device
  - /lib/modules: volume encryption by `cryptsetup`
  - /proc: To access host's mount namespace by `nsenter`
  - /sys: Volume encryption by `cryptsetup`
- Longhorn CSI Plugin

>   - /: For RWX shared volume

Reviewer 2: For this volume mount, here is the more detailed explanation: https://github.com/longhorn/longhorn-manager/blob/8177316cafe002b512577bb4f05a2065e80ed383/csi/node_server.go#L610

Author: I thought the reason is we need to support RWX volume https://github.com/longhorn/longhorn-manager/commit/a644b94439754eb791cf31955ec8b8bff8abf30b.

But if we want to grant permission to `nsenter`, shouldn't we just need to allow `/proc` mount point rather than the root mount point?

Reviewer 2: Yeah, I am confused about this as well. The mounter is actually [this script](https://github.com/longhorn/longhorn-manager/blob/master/package/nsmounter), in which we use `-t 1` rather than something like `--net=/host/proc/1/ns/net --mount=/host/proc/1/ns/mnt`. @joshimoo Can you provide more details?

Reviewer 3: @jenting @shuo-wu We no longer need the `/rootfs` since I have removed the nfs mounter and replaced it with the nsenter script above. Since we run the csi plugin with `hostPid: true` we can utilize `-t 1` since `/proc` maps to `/proc` in this case.

It's a left over from before where we were doing host checks via the custom mounter.

Author: @joshimoo So we could remove these lines https://github.com/longhorn/longhorn-manager/blob/1664880ddaedb9875671714b001a2ad7c9bc3b5c/csi/deployment.go#L525-L532 and https://github.com/longhorn/longhorn-manager/blob/1664880ddaedb9875671714b001a2ad7c9bc3b5c/csi/deployment.go#L459-L463, right?

Reviewer 2: Then we can mention that this will be removed since v1.2.4.

Author: I'd prefer to add the wording that this will be removed in the coming release. 

I'm not sure if we can get this [improvement](https://github.com/longhorn/longhorn/issues/3736) into v1.2.4.
cc @innobead

Reviewer: I don't think we should at this moment if it's not in 1.2.4 yet.

Reviewer 2: Sure. Then we don't need to mention in which specific version we will remove it.

>   - /: Access any data path on this node and access Longhorn engine binaries.

Reviewer: Saw it requires the root path, but what cases actually happening in IM-R pods use that path?

Author: I think it simplifies the customize path for storing volume data on a host.

Below is the command to launch a replica process with the specific path to storing the replica data.
```shell
/host/var/lib/longhorn/engine-binaries/rancher-mirrored-longhornio-longhorn-engine-v1.1.1/longhorn replica /host/var/lib/longhorn/replicas/pvc-3a27bbc4-bd70-4a93-8aca-154ff88df514-9ead319e --size 1073741824 --listen 0.0.0.0:10000
```

But the better approach _probably_ mounts a different hostpath if the users customize the longhorn storage path rather than the root mount point.

>   - /dev: Longhorn is a block storage solution for Kubernetes. The block device is created under `/dev` path.

Reviewer: ```suggestion
  - /dev: Block devices created by Longhorn are under the `/dev` path.
```

>   - /lib/modules: Kernel module requried by `cryptsetup`.

Reviewer: ```suggestion
  - /lib/modules: Kernel modules requried by `cryptsetup` for volume encryption.
```

>   - /proc: To access host's mount namespace by `nsenter` across different namepsaces.

Reviewer: ```suggestion
  - /proc: Find the recognized host process like container runtime, then use `nsenter` for volume encryption.
```

  - /dev: Volume block device
  - /lib/modules: Kubernetes CSI plugin
  - /sys: Volume encryption by `cryptsetup`
  - /var/lib/kubelet/plugins/kubernetes.io/csi: Kubernetes CSI plugin
  - /var/lib/kubelet/plugins_registry: Kubernetes CSI plugin
  - /var/lib/kubelet/plugins/driver.longhorn.io: Kubernetes CSI plugin

>   - /var/lib/kubelet/pods: Kubernetes CSI plugin

Reviewer 2: Not sure if we need to explain this part a little bit more.

Reviewer: Probably we should.  The intention is to let user understand why the CSI compatible storage solution like longhorn need these privileged permissions besides our components. However, the general explanation is good enough.

>   - /sys: To support volume encryption by `cryptsetup`.

Reviewer: ```suggestion
  - /sys: Support volume encryption by `cryptsetup`.
```

>   - /: For RWX shared volume.

Reviewer 3: "/" should no longer be needed with the new nsenter shell script and the removed nfsmounter.
Since we are running with `hostPid: true` -> `/proc => /proc`

Author: Since within v1.2.3 Longhorn code, we still have mount point root mount point inside the CSI plugins.
Therefore, we should list the root mount point in the doc to let the user/customer knows the purpose from the security point of view.

>   - /dev: Longhorn is a block storage solution for Kubernetes. The block device is created under `/dev` path.

Reviewer: ```suggestion
  - /dev: Block devices created by Longhorn are under the `/dev` path.
```

>   - /lib/modules: Kernel module required by Kubernetes CSI plugin.

Reviewer: ```suggestion
  - /lib/modules: Kernel modules required by Longhorn CSI plugin.
```

>   - /sys: To support volume encryption by `cryptsetup`.

Reviewer 2: For the CSI plugin, this mount is introduced before implementing the volume encryption. Is there any other reason for it?

Reviewer 2: I am not familiar with it, either. Not sure if others know anything about this.

Reviewer: yeah, we can figure out it later. 

cc @derekbit @joshimoo

>   - /sys: To support volume encryption by `cryptsetup`.

Reviewer: ```suggestion
  - /sys: Support volume encryption by `cryptsetup`.
```

>   - /var/lib/kubelet/plugins/kubernetes.io/csi: The path where the Kubernetes CSI plugin creates a staging path (via `NodeStageVolume`) of a block device. Then, Kubernetes CSI plugin bind mount the staging path to target path `/var/lib/kubelet/pods` (via `NodePublishVolume`) for support single volume could be mounted to multiple Pods.

Reviewer: ```suggestion
  - /var/lib/kubelet/plugins/kubernetes.io/csi: The path where the Longhorn CSI plugin creates the staging path (via `NodeStageVolume`) of a block device. The staging path will be bind-mounted to the target path `/var/lib/kubelet/pods` (via `NodePublishVolume`) for support single volume could be mounted to multiple Pods.
```

>   - /var/lib/kubelet/plugins/driver.longhorn.io: The path where the socket for kubelet -> CSI driver communication is done.

Reviewer: ```suggestion
  - /var/lib/kubelet/plugins/driver.longhorn.io: The path where the socket for the communication between kubelet Longhorn CSI driver.
```

>   - /var/lib/kubelet/pods: The path where the CSI driver mounts volume from staging to target path (via `NodePublishVolume`).

Reviewer: ```suggestion
  - /var/lib/kubelet/pods: The path where the Longhorn CSI driver mounts volume from the target path (via `NodePublishVolume`).
```

- Longhorn CSI Attacher/Provisioner/Resizer/Snapshotter

>   - /var/lib/kubelet/plugins/driver.longhorn.io: The path where the socket for kubelet -> CSI driver communication is done.

Reviewer: ```suggestion
  - /var/lib/kubelet/plugins/driver.longhorn.io: The path where the socket for the communication between kubelet Longhorn CSI driver.
```

- Longhorn Backing Image Manager

>   - /var/lib/longhorn (the default data path, could be customized by the user): The path to use for storing data on a host.

Reviewer: ```suggestion
  - /var/lib/longhorn: The default path for storing data on the host.
```

- Longhorn Backing Image Data Source

>   - /var/lib/longhorn (the default data path, could be customized by the user): The path to use for storing data on a host.

Reviewer: ```suggestion
  - /var/lib/longhorn: The default path for storing data on the host.
```

### Root and Privileged Permission

> The longhorn components need to access the host's disk in the container.

Reviewer: ```suggestion
Longhorn components require root access with privileged permissions to achieve volume operations and management, because Longhorn relies on system resources management on the host across different namespaces, for example, Longhorn uses `nsenter` to understand block devices' usage or encrypt/decrypt volumes on the host.
```

> A use case is to access disk metadata and replicas directory on the host. The list of the directories of all components are:

Reviewer: Below are examples to list the different use cases requiring root access and privileged permissions.

- Longhorn Manager

>   - /dev: Volume block device

Reviewer: Suggest having an explanatory brief statement to say what we need to access `/dev` instead of just a noun. The same as others.

- Longhorn Instance Manager (Engine)
  - /dev: Volume block device
  - /proc: To access host's mount namespace
  - /var/lib/longhorn/engine-binaries: Longhorn engine binary
- Longhorn Instance Manager (Replica)
  - /: Data path (allow the user to customize data path)
- Longhorn Share Manager
  - /dev: Volume block device
  - /lib/modules: volume encryption by `cryptsetup`
  - /proc: To access host's mount namespace by `nsenter`
  - /sys: Volume encryption by `cryptsetup`
- Longhorn CSI Plugin
  - /: For RWX shared volume
  - /dev: Volume block device
  - /lib/modules: Kubernetes CSI plugin
  - /sys: Volume encryption by `cryptsetup`
  - /var/lib/kubelet/plugins/kubernetes.io/csi: Kubernetes CSI plugin
  - /var/lib/kubelet/plugins_registry: Kubernetes CSI plugin
  - /var/lib/kubelet/plugins/driver.longhorn.io: Kubernetes CSI plugin
  - /var/lib/kubelet/pods: Kubernetes CSI plugin
- Longhorn CSI Attacher/Provisioner/Resizer/Snapshotter
  - /var/lib/kubelet/plugins/driver.longhorn.io: Kubernetes CSI plugin
- Longhorn Backing Image Manager
  - /var/lib/longhorn: Data path
- Longhorn Backing Image Data Source
  - /var/lib/longhorn: Data path

> Besides, Longhorn uses `nsenter` command to execute commands on the host. The `nsenter` command, for example, is to check the host's disk space usage

Reviewer: We can remove this part, but move to the summary instead.
